File: backend/test/com/skewwhiffy/auraltester/abc/SingleLineAbcTest.scala
package com.skewwhiffy.auraltester.abc

import com.skewwhiffy.auraltester.clefs.{Clef, ClefFactory}
import com.skewwhiffy.auraltester.internalnotation.NoteFactory
import com.skewwhiffy.auraltester.notes.{AbsoluteNote, NoteLength}
import com.skewwhiffy.auraltester.scales.Key
import com.skewwhiffy.auraltester.testutils.{MockInstantiation, TestData}
import org.scalatest.{BeforeAndAfter, GivenWhenThen}
import org.scalatest.flatspec.AnyFlatSpec

class SingleLineAbcTest extends AnyFlatSpec with MockInstantiation with GivenWhenThen with BeforeAndAfter {
  private var title: String = _
  private var clef: Clef = _
  private var noteLength: NoteLength = _
  private var notes: List[AbsoluteNote] = _
  private var notesAbc: List[String] = _

  before {
    val noteFactory = new NoteFactory()
    val clefFactory = new ClefFactory(noteFactory)
    title = TestData.random.string
    clef = TestData.random.oneOf(clefFactory.treble, clefFactory.alto, clefFactory.tenor, clefFactory.bass)
    noteLength = TestData.random.oneOf(
      NoteLength.breve,
      NoteLength.semibreve,
      NoteLength.minim,
      NoteLength.crotchet,
      NoteLength.quaver
    )
    notesAbc = Range(0, 10).map(_ => TestData.random.string).toList
    notes = notesAbc.map(it => {
      val note = mock[AbsoluteNote]
      when(note.abc(Key.cMajor)).thenReturn(it)
      note
    })
  }

  it should "populate title" in {
    Given("title is supplied")

    When("getting single line ABC")
    val abc = SingleLineAbc(title, clef, noteLength, List(notes))

    Then("abc has index")
    assert(abc.abc.contains("X:1"))
    And("populates title")
    assert(abc.abc.contains(s"T:$title"))
    And("clef is populated")
    assert(abc.abc.contains(s"K:clef=${clef.abc}"))
    And("note length is populated")
    assert(abc.abc.contains(s"L:${noteLength.abc}"))
    And("notes abc is correct")
    assert(abc.abc.contains(notesAbc.mkString))
  }

  it should "not populate title" in {
    Given("title is not supplied")

    When("getting single line ABC")
    val abc = SingleLineAbc(clef, noteLength, List(notes))

    Then("abc has no title")
    assert(!abc.abc.contains("T:"))
  }

  // TODO: Multiple bars
}

File: backend/src/test/kotlin/com/skewwhiffy/notation/model/abc/SingleLineAbcTest.kt
package com.skewwhiffy.notation.model.abc

import com.skewwhiffy.notation.factory.ClefFactory
import com.skewwhiffy.notation.factory.NoteFactory
import com.skewwhiffy.notation.model.AbsoluteNote
import com.skewwhiffy.notation.model.Clef
import com.skewwhiffy.notation.model.Key
import com.skewwhiffy.notation.model.NoteLength
import com.skewwhiffy.test.util.TestData
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.File

class SingleLineAbcTest {
  private lateinit var title: String
  private lateinit var clef: Clef
  private lateinit var noteLength: NoteLength
  private lateinit var notes: List<AbsoluteNote>
  private lateinit var notesAbc: List<String>

  @BeforeEach
  fun `before each`() {
    val noteFactory = NoteFactory()
    val clefFactory = ClefFactory(noteFactory)
    title = TestData.random.string
    clef = TestData.random.oneOf(clefFactory.treble, clefFactory.alto, clefFactory.tenor, clefFactory.bass)
    noteLength = TestData.random.oneOf(
      NoteLength.breve,
      NoteLength.semibreve,
      NoteLength.minim,
      NoteLength.crotchet,
      NoteLength.quaver
    )
    notes = (1..10).map { TestData.random.absoluteNote }
    notesAbc = notes.map { it.abc(Key.cMajor) }
  }

  @Test
  fun `populates title`() {
    val abc = SingleLineAbc(title, clef, noteLength, listOf(notes))

    assertThat(abc.abc).contains("X:1")
    assertThat(abc.abc).contains("T:$title")
    assertThat(abc.abc).contains("K:clef=${clef.abc}")
    assertThat(abc.abc).contains("L:${noteLength.abc}")
    assertThat(abc.abc).contains(notesAbc.joinToString(""))
  }

  @Test
  fun `not populate title`() {
    val abc = SingleLineAbc(clef, noteLength, listOf(notes))

    assertThat(abc.abc).doesNotContain("T:")
  }

  // TODO: Multiple bars
}